sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ImplementISerializableCorrectly.cs

```diff
@@ -0,0 +1,246 @@
+/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2017 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public class ImplementISerializableCorrectly : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S3925";
+        internal const string MessageFormat = "Update this implementation of 'ISerializable' to conform to the recommended serialization pattern.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                if (c.SemanticModel.Compilation.IsTest())
+                {
+                    return;
+                }
+
+                var classDeclaration = (ClassDeclarationSyntax)c.Node;
+                var typeSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
+
+                if (!ImplementsISerializable(typeSymbol))
+                {
+                    return;
+                }
+
+                var getObjectData = typeSymbol.GetMembers()
+                    .OfType<IMethodSymbol>()
+                    .FirstOrDefault(KnownMethods.IsGetObjectData);
+
+                var implementationErrors = new List<SecondaryLocation>();
+
+                implementationErrors.AddRange(CheckSerializableAttribute(classDeclaration, typeSymbol));

```

Comment:
Not sure if it is actually better or not but maybe you could use `Concat` instead of `AddRange`